ttnn/cpp/ttnn/operations/reduction/topk/device/kernels/compute/topk_final.cpp

```diff
@@ -0,0 +1,166 @@
+
+
+        // iterative divide and conquer on pairs of tiles (bitonic topk merge and rebuild)
+        // first iteration we compare 0th and 1st tile, then 2nd and 3rd, etc. We get the sorted top 32 values in each pair.
+        // second iteration we compare 0th and 2nd tile, then 4th and 6th, etc.
+        // logWt iteration we compare 0th and Wt/2 tile
+        // single buffer as we can pack tiles back in-place
+        for (uint32_t m_iter = 0; m_iter < logWt; ++m_iter) {
+            bool a = false;
+            cb_wait_front(input_transposed_cb_index, Wt);
+            cb_wait_front(index_transposed_cb_index, Wt);
+
+            for (uint32_t left_ind = 0; left_ind < Wt - (1 << m_iter); left_ind += 2 << m_iter) {
+                uint32_t right_ind = left_ind + (1 << m_iter);
+                acquire_dst(tt::DstMode::Half);
+
+                // unpack values into dest
+                copy_tile_to_dst_init_short_with_dt(index_transposed_cb_index, input_transposed_cb_index);
+                copy_tile(input_transposed_cb_index, left_ind, input_dest_start);
+                copy_tile(input_transposed_cb_index, right_ind, input_dest_end);
+
+                // unpack indices into dest
+                copy_tile_to_dst_init_short_with_dt(input_transposed_cb_index, index_transposed_cb_index);
+                copy_tile(index_transposed_cb_index, left_ind, index_dest_start);
+                copy_tile(index_transposed_cb_index, right_ind, index_dest_end);
+
+                // merge values - move larger 32 values into 0th dest and lower 32 values into 1st dest
+                ckernel::topk_merge(0, m_iter, K);
+                // sort within the larger 32 values
+                ckernel::topk_rebuild(0, (uint32_t) a, m_iter, K, logk, true);
+
+                // pack value tiles in-place in the single-buffered cb_intermed0, we only need the upper 32 values for topk, which was in input_dest_start
+                pack_reconfig_data_format(input_transposed_cb_index);
+                pack_tile<true>(input_dest_start, input_transposed_cb_index, left_ind);
+
+                // pack index tiles in-place in the single-buffered cb_intermed1, we only need the upper 32 values for topk, which was in index_dest_start
+                pack_reconfig_data_format(index_transposed_cb_index);
+                pack_tile<true>(index_dest_start, index_transposed_cb_index, left_ind);
+                release_dst(tt::DstMode::Half);
+                a = !a;
+            }
+            cb_reserve_back(input_transposed_cb_index, Wt);
+            cb_reserve_back(index_transposed_cb_index, Wt);
+
+            cb_pop_front(input_transposed_cb_index, Wt);
+            cb_pop_front(index_transposed_cb_index, Wt);
+
+            cb_push_back(input_transposed_cb_index, Wt);
+            cb_push_back(index_transposed_cb_index, Wt);
+        }
+
+        constexpr uint32_t Kt =  K % TILE_WIDTH == 0 ? K/TILE_WIDTH : K/TILE_WIDTH + 1;
```

Comment:
This looks like a `divup` and should be done in host code. I don't think `K` is used anywhere?